### ISSUE TITLE
Typegen Usage which worked with 9.14.2 doesn't work with 10.1.2

### ISSUE THREAD

**Issue by rgoulter** [polkadot-js/api#5540]
* **I'm submitting a ...**

  - [ ] Bug report
  - [ ] Feature request
  - [X] Support request
  - [ ] Other

* **What is the current behavior and expected behavior?**

Code which worked with `@polkadot/api` `v9.14.2` doesn't work with `v10.1.2`. -- Likely same as https://github.com/polkadot-js/api/issues/5524#issuecomment-1468321718

Some clarification on what changes are needed would be useful.

Example code (& the commands used) can be found here: https://github.com/rgoulter/polkadot-js-typegen-efinity -- It's a pretty bare-bones usage of `typegen`, to generate the code, and includes a small `example` TypeScript program which uses the generated types.

The latest commit there updates the `@polkadot/typegen` to `10.1.2` (and in `example/package.json`, its `@polkadot/api` to `10.1.2`). 

I'm having trouble figuring out what changes I need to make to get it to work with `10.1.2`.

### What goes wrong:

When it runs `polkadot-types-from-defs`, it has problems loading the `definitions.ts`:

```
> ts-node --skip-project node_modules/.bin/polkadot-types-from-defs --package @efinity/api-augment/interfaces --input ./src/interfaces --endpoint ./metadata.json

ERROR: Unable to load user definitions: Unknown file extension ".ts" for /home/rgoulter/scratch/efinity-api-augment/src/interfaces/definitions.ts
```

When running `polkadot-types-from-chain`, some `.ts` code is generated, but the generated code then leads to errors when trying to build it.

```
> ts-node --skip-project node_modules/.bin/polkadot-types-from-chain --endpoint ./metadata.json --output ./src/interfaces

Generating from metadata, 254,086 bytes
ERROR: No lookup definitions found: Unknown file extension ".ts" for /home/rgoulter/scratch/efinity-api-augment/src/interfaces/lookup.ts
2023-03-15 14:41:52        REGISTRY: Unknown signed extensions CheckFuelTank found, treating them as no-effect
/home/rgoulter/scratch/efinity-api-augment/src/interfaces/augment-api-consts.ts
        Generating
        Writing

...
```

Sample errors when trying to run `tsc` in `example`:

```
node_modules/@efinity/api-augment/src/interfaces/augment-api-consts.ts:166:17 - error TS2304: Cannot find name 'FrameSupportPalletId'.

166       palletId: FrameSupportPalletId & AugmentedConst<ApiType>;
                    ~~~~~~~~~~~~~~~~~~~~

...

Found 457 errors in 4 files.

Errors  Files
    11  node_modules/@efinity/api-augment/src/interfaces/augment-api-consts.ts:166
   142  node_modules/@efinity/api-augment/src/interfaces/augment-api-events.ts:43
    86  node_modules/@efinity/api-augment/src/interfaces/augment-api-query.ts:31
   218  node_modules/@efinity/api-augment/src/interfaces/augment-api-tx.ts:23
```

### Problems running with `ts-node-esm`, too:

If instead of `ts-node`, I try `ts-node-esm`, `polkadot-types-from-defs` abruptly exits without an error message:

```
> ts-node-esm --skip-project node_modules/.bin/polkadot-types-from-defs --package @efinity/api-augment/interfaces --input ./src/interfaces --endpoint ./metadata.json

/home/rgoulter/scratch/efinity-api-augment/src/interfaces/types.ts
        Generating
        Extracting interfaces for default
```

If instead of `ts-node`, I try `ts-node-esm`, `polkadot-types-from-chain` is still unable to lookup the definitions:

```
> @efinity/api-augment@0.0.1 generate:meta
> ts-node-esm --skip-project node_modules/.bin/polkadot-types-from-chain --endpoint ./metadata.json --output ./src/interfaces

Generating from metadata, 254,086 bytes
ERROR: No lookup definitions found: Cannot read properties of undefined (reading 'default')
2023-03-15 14:45:04        REGISTRY: Unknown signed extensions CheckFuelTank found, treating them as no-effect
/
...
```

* **What is the motivation for changing the behavior?**

n/a

* **Please tell us about your environment:**

Linux.

```
$ node --version
v18.12.1

$ npx tsc --version
Version 4.9.4
```

**Comment by jacogr** [comment]
Ok, I found pages and pages of threads on google with regards to this. So at least we are not the first...

The quick fix seems to be (which mostly works for me on your reproduction) -

- Add `"type": "module"` to your `package.json`
- Add `--esm` to the `ts-node` flags, i.e `ts-node --esm --skip-project node_modules/.bin/...`

(In your specific example there is `ERROR: No lookup definitions found: exports is not defined in ES module scope` - which is due to the empty `exports {}` in `defintions.ts`)

What is weird here is that you would expect, when running as ESM, you would not need `--esm` flags and change the module type for the generator part. So basically it would seem that `10.x` is more breaking that planned as a major.

TL;DR the above works on the sample for me - I'm not sure if there are other solutions that can smooth things out more.

**Comment by rgoulter** [comment]
> ... Doing that makes it generate everything. ...

Part of the trouble is that, if `polkadot-types-from-chain` shows some `ERROR` related to lookup definitions, the generated TS code isn't usable.

e.g. when there's some error related to lookup definitions, the generated `augment-api-consts.ts` is:

```
import '@polkadot/api-base/types/consts';

import type { ApiTypes, AugmentedConst } from '@polkadot/api-base/types';
import type { Bytes, Option, U8aFixed, bool, u128, u16, u32, u64, u8 } from '@polkadot/types-codec';
import type { Codec } from '@polkadot/types-codec/types';
import type { AccountId32, Perbill, Permill } from '@polkadot/types/interfaces/runtime';

export type __AugmentedConst<ApiType extends ApiTypes> = AugmentedConst<ApiType>;

declare module '@polkadot/api-base/types/consts' {
```

whereas it 'should' look more like:

```
import '@polkadot/api-base/types/consts';

import type { ApiTypes, AugmentedConst } from '@polkadot/api-base/types';
import type { Bytes, Option, U8aFixed, bool, u128, u16, u32, u64, u8 } from '@polkadot/types-codec';
import type { Codec } from '@polkadot/types-codec/types';
import type { AccountId32, Perbill, Permill } from '@polkadot/types/interfaces/runtime';
import type {
    ...
    FrameSupportPalletId,
    ...
} from '@polkadot/types/lookup';

export type __AugmentedConst<ApiType extends ApiTypes> = AugmentedConst<ApiType>;

declare module '@polkadot/api-base/types/consts' {
```

And because of the missing `import { ... } from @polkadot/types/lookup`, the code which relies on it breaks.

- - -

Rather, in the example, the problem isn't so much generating the code, so much as `polkadot-types-from-chain` being able to read the lookup types properly.

(In my example, I can ignore `polkadot-types-from-defs` being unable to load `definitions.ts`; but it does seem untidy to ignore the `ERROR`).

I can kinda-sorta work around it by directly modifying the `node_modules/@polkadot/typegen/fromChain.js`

```
    try {
        const lookPath = assertFile(path.join(outputPath, 'lookup.ts'));
        const lookCont = await import(lookPath);
        customLookupDefinitions = {
            rpc: {},
            types: lookCont.module.default
        };
    }
```

to have `lookCont.default.default` instead of `lookCont.default.default`; but, I can't find out "why" the result of `await import(...)` would have `.module` or `.default`.

**Comment by jacogr** [comment]
`.module` should not be there () PR incoming.

`.default` will be there for .e.g lookup where it does an `export default ...`

**Comment by montogeek** [comment]
I got this error:
<img width="381" alt="image" src="https://user-images.githubusercontent.com/1002461/225296467-00193a3a-18a7-4cc0-9d3f-c7c3fc1d7acc.png">

With a `definitions.ts` like:
```
export { default as identity } from './identity/definitions';
```

Logging `usersDefs` shows:

```
[Module: null prototype] { default: { identity: { types: [Object] } } }
```

So I changed it back to `userDefs = defCont.default;`

It works for a bit, but then at the end throws this error:
```
disconnected, code: '1005' reason: ''
```

Logging the error, a property says that `wasClean: true`, but the good news is that I get the expected types.

**Comment by montogeek** [comment]
Another new problem in 10.1.2 and 10.1.3 is that `types.ts` is now exporting from a JS file:
In 9.10.5
```
export * from './identity/types';
```

In 10.1.2 and 10.1.3
```
export * from './identity/types.js';
```

This breaks the build, I am using the `--esm` option, will try with a different loader.

**Comment by jacogr** [comment]
That should really not break the build at all. It work in both TS `moduleResultion: node` and `moduleResolution: nodenext` (which is why it has been added)

**Comment by jacogr** [comment]
@montogeek On your `export { default as identity } from './identity/definitions';` issue - I actually added a couple of those in my tests as well, there is no default export for me on those. (Logging the keys of the object should show ['identity], ...]`)

**Comment by montogeek** [comment]
What Node version are you using?

**Comment by jacogr** [comment]
`v18.14.0`

**Comment by montogeek** [comment]
I am using Node 16.16, still get the same problem when using `18.14.0`.

My command is:
```
ts-node --esm --skipProject ./node_modules/@polkadot/typegen/scripts/polkadot-types-from-defs.mjs --package . --input ./libs/chain-types/src/lib/interfaces/ --endpoint wss://url
```

I do not use the script directly because of this error:
```
❯ node node_modules/.bin/polkadot-types-from-defs --package . --input ./libs/chain-types/src/lib/interfaces/ --endpoint wss://tee-staging.litentry.io                       ─╯
/Users/fernando/Code/project/node_modules/.bin/polkadot-types-from-defs:2
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^

SyntaxError: missing ) after argument list
    at internalCompileFunction (node:internal/vm:73:18)
    at wrapSafe (node:internal/modules/cjs/loader:1149:20)
    at Module._compile (node:internal/modules/cjs/loader:1190:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1280:10)
    at Module.load (node:internal/modules/cjs/loader:1089:32)
    at Module._load (node:internal/modules/cjs/loader:930:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47

Node.js v18.14.0
```

**Comment by jacogr** [comment]
Something is weird here. See the 2 commands and the outputs -

```bash
> ts-node --esm --skipProject ./node_modules/@polkadot/typegen/scripts/polkadot-types-from-defs.mjs --package @efinity/api-augment/interfaces --input ./src/interfaces --endpoint ./metadata.json

ERROR: Unable to load user definitions: exports is not defined in ES module scope
<somewhere>/polkadot-js-typegen-efinity/src/interfaces/types.ts
	Generating
	Writing

<somewhere>/polkadot-js-typegen-efinity/src/interfaces/augment-types.ts
	Generating
	Writing

2023-03-15 14:42:58        REGISTRY: Unknown signed extensions CheckFuelTank found, treating them as no-effect
<somewhere>/polkadot-js-typegen-efinity/src/interfaces/lookup.ts
	Generating
	Writing
```

And then via loader (no `exports` message link above) -

```bash
> node --loader ts-node/esm ./node_modules/@polkadot/typegen/scripts/polkadot-types-from-defs.mjs --package @efinity/api-augment/interfaces --input ./src/interfaces --endpoint ./metadata.json

(node:27028) ExperimentalWarning: Custom ESM Loaders is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
/Users/jacogr/Projects/polkadotjs/test/polkadot-js-typegen-efinity/src/interfaces/types.ts
	Generating
	Extracting interfaces for something
	Writing

/Users/jacogr/Projects/polkadotjs/test/polkadot-js-typegen-efinity/src/interfaces/augment-types.ts
	Generating
	Writing

2023-03-15 14:43:31        REGISTRY: Unknown signed extensions CheckFuelTank found, treating them as no-effect
/Users/jacogr/Projects/polkadotjs/test/polkadot-js-typegen-efinity/src/interfaces/lookup.ts
	Generating
	Writing
```

Have no clue what the difference would be between `ts-node --esm ...` and `node --loader ts-node/esm`

**Comment by jacogr** [comment]
I'm not convinced that `ts-node --esm` actually works as expected. My outputs in https://github.com/polkadot-js/api/issues/5540#issuecomment-1469943748 above matches what you are seeing, however...

```bash
node --loader ts-node/esm node_modules/.bin/polkadot-types-from-defs --package @moonbeam/...
```

at least above did the trick.

**Comment by gdethier** [comment]
@jacogr By applying the option 1 described [here](https://github.com/polkadot-js/docs/issues/419), I am indeed able to generate definitions. However, chain types are not generated (`polkadot-types-from-chain`), all I get is this:

```
(node:41574) ExperimentalWarning: Custom ESM Loaders is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
connected
Generating from metadata, 105,931 bytes
disconnected, code: '1005' reason: ''
```

After a couple of hours debugging, I figured out that the process is actually killed because of the 1005 event code (see [this line](https://github.com/polkadot-js/api/blob/3b39de3852aa79ecba5179acc3dbf9b59c077ebd/packages/typegen/src/util/wsMeta.ts#L20)) and type generation is simply not executed.

A workaround is to comment out the line. It does not solve the unexpected event code issue, but at least types are properly generated.

**Comment by jacogr** [comment]
Indeed. Looking at the above code link promise is actually not properly tracked/handled -

- it does resolve when a response is received (all good)
- it does not reject onerror, rather exits (same with close)

So basically that should be adjusted to only resolve/reject once (onmessage/onerror/onclose) aka when we have a response, we don't care if things go wrong with the connection so should not perform any exit inside there. Said differently whichever of there we find once should determine the Promise result... once.

(It seems it was expecting all RPCs nodes to always behave and that assumption doesn't always hold)

**Comment by MrishoLukamba** [comment]
Am still getting errors in lookup.ts as it is not generated as ES module I think and cannot export the types to other modules hence breaking the code.
@jacogr 
![image](https://user-images.githubusercontent.com/69342343/235455548-130718be-8f2e-48d2-8b8c-6f9751f01ea8.png)

**Comment by Verin1005** [comment]
In version 10.1.2, the export method has changed：
`export * from './identity/types.js';` 

in version 9.14.2:
`export * from './identity/types';`

 which disrupts the structure. I have tried all the mentioned methods above, but none of them successfully resolve the issue.

 https://github.com/litentry/litentry-parachain/pull/1703

**Comment by devdanco** [comment]
I have created PR for [docs](https://github.com/polkadot-js/docs/pull/445) on how we generate these types. I hope that it will be beneficial to someone.
